primers.go

```diff
@@ -103,3 +104,182 @@ func MeltingTemp(sequence string) float64 {
+    called ligase, which basically just glues together DNA fragments. Wikipedia
+    has a good introduction:
+    https://en.wikipedia.org/wiki/Ligation_(molecular_biology)
+
+******************************************************************************/
+
+// NucleobaseDeBruijnSequence generates a DNA DeBruijn sequence with alphabet ATGC. DeBruijn sequences are basically a string with all unique substrings of an alphabet represented exactly once. Code is adapted from https://rosettacode.org/wiki/De_Bruijn_sequences#Go
+func NucleobaseDeBruijnSequence(substringLength int) string {
+	alphabet := "ATGC"
+	alphabetLength := len(alphabet)
+	a := make([]byte, alphabetLength*substringLength)
+	var seq []byte
+	// The following function is mainly adapted from rosettacode.
+	var ConstructDeBruijn func(int, int) // recursive closure
+	ConstructDeBruijn = func(t, p int) {
+		if t > substringLength {
+			if substringLength%p == 0 {
+				seq = append(seq, a[1:p+1]...)
+			}
+		} else {
+			a[t] = a[t-p]
+			ConstructDeBruijn(t+1, p)
+			for j := int(a[t-p] + 1); j < alphabetLength; j++ {
+				a[t] = byte(j)
+				ConstructDeBruijn(t+1, t)
+			}
+		}
+	}
+	ConstructDeBruijn(1, 1)
+	var buf bytes.Buffer
+	for _, i := range seq {
+		buf.WriteByte(alphabet[i])
+	}
+	b := buf.String()
+	return b + b[0:substringLength-1] // as cyclic append first (n-1) digits
+}
+
+// CreateBarcodesWithBannedSequences creates a list of barcodes given a desired barcode length, the maxSubSequence shared in each barcode,
+// Sequences may be marked as banned by passing a static list, `bannedSequences`, or, if more flexibility is needed, through a list of `bannedFunctions` that dynamically generates bannedSequences.
+// If a sequence is banned, it will not appear within a barcode. The a `bannedFunctions` function can determine if a barcode should be banned or not on the fly. If it is banned, we will continuing iterating until a barcode is found that satisfies the bannedFunction requirement.
+func CreateBarcodesWithBannedSequences(length int, maxSubSequence int, bannedSequences []string, bannedFunctions []func(string) bool) []string {
+	var barcodes []string
+	var start int
+	var end int
+	debruijn := NucleobaseDeBruijnSequence(maxSubSequence)
+	for barcodeNum := 0; (barcodeNum*(length-(maxSubSequence-1)))+length < len(debruijn); {
+		start = barcodeNum * (length - (maxSubSequence - 1))
+		end = start + length
+		barcodeNum++
+		for _, bannedSequence := range bannedSequences {
+			// If the current deBruijn range has the banned sequence, iterate one base pair ahead. If the iteration reaches the end of the deBruijn sequence, close the channel and return the function.
+			for strings.Contains(debruijn[start:end], bannedSequence) {
+				if end+1 > len(debruijn) {
+					return barcodes
+				}
+				start++
+				end++
+				barcodeNum++
+			}
+			// Check reverse complement as well for the banned sequence
+			for strings.Contains(debruijn[start:end], ReverseComplement(bannedSequence)) {
```

Comment:
This code block for checking for bannedsequences in reverse complements is untested. Test it and I think we got a merge here!